mqtt/mqtt-broker/benches/dispatch_messages.rs

```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    async fn run(mut self) {
+        let mut stop = false;
+        let mut messages_received = 0u32;
+        while let Some(message) = self.rx.recv().await {
+            match message {
+                Message::Client(client_id, event) => {
+                    let event_out = match event {
+                        ClientEvent::PublishTo(Publish::QoS0(id, publish))
+                        | ClientEvent::PublishTo(Publish::QoS12(id, publish)) => {
+                            messages_received += 1;
+
+                            match publish.packet_identifier_dup_qos {
+                                proto::PacketIdentifierDupQoS::AtMostOnce => {
+                                    Some(ClientEvent::PubAck0(id))
+                                }
+                                proto::PacketIdentifierDupQoS::AtLeastOnce(id, _) => {
+                                    let mut acks =
+                                        self.pubs_to_be_acked.lock().expect("pubs_to_be_acked");
+                                    acks.insert(id);
+                                    Some(ClientEvent::PubAck(proto::PubAck {
+                                        packet_identifier: id,
+                                    }))
+                                }
+                                proto::PacketIdentifierDupQoS::ExactlyOnce(id, _) => {
+                                    Some(ClientEvent::PubRec(proto::PubRec {
+                                        packet_identifier: id,
+                                    }))
+                                }
+                            }
+                        }
+                        ClientEvent::PublishFrom(_) => None,
+                        ClientEvent::PubAck0(_) => None,
+                        ClientEvent::PubAck(puback) => {
+                            let mut acks = self.pubs_to_be_acked.lock().expect("pubs_to_be_acked");
+                            if !acks.remove(&puback.packet_identifier) {
+                                warn!(
+                                    "{}: cannot find packet identifier {}",
+                                    self.client_id, puback.packet_identifier,
+                                );
+                            }
+                            None
+                        }
+                        ClientEvent::PubRec(_) => None,
+                        ClientEvent::PubRel(_) => None,
+                        ClientEvent::PubComp(_) => None,
+                        ClientEvent::DropConnection
+                        | ClientEvent::CloseSession
+                        | ClientEvent::Disconnect(_) => {
+                            stop = true;
+                            None
+                        }
+                        ClientEvent::PingReq(_) => Some(ClientEvent::PingResp(proto::PingResp)),
+                        ClientEvent::PingResp(_) => Some(ClientEvent::PingReq(proto::PingReq)),
```

Comment:
Maybe I'm missing something, but are you sure you need to reply with PingReq on PingResp? won't you get an "infinite loop" of pings?

Reply:
The idea was the client responds to the broker ping request. But I wonder if the broker actively checks clients. Probably not...